### ISSUE TITLE
aws_spot_instance_request with associate_public_ip_address=true never assign public ip.

### ISSUE THREAD

**Issue by maskinoshita** [hashicorp/terraform#9575]
I'm trying to launch a spot instance with public ip in VPC. My EIPs are reached limit, so I need to use auto-assigned public ip. I set associate_public_ip_address true, but public ip is never assigned. I also tried an ondemand instance. In this case, a public ip was assigned.

vpc(subnet) setting:
- Auto-assign Public IP: no

According to aws cli doc, spot instance with public ip can be done with followed code.
http://docs.aws.amazon.com/cli/latest/reference/ec2/request-spot-instances.html#examples

command:

``` bash
aws ec2 request-spot-instances --spot-price "0.050" --instance-count 1 --type "one-time" --launch-specification file://specification.json
```

spcification.json:

``` json
{
  "ImageId": "ami-1a2b3c4d",
  "KeyName": "my-key-pair",
  "InstanceType": "m3.medium",
  "NetworkInterfaces": [
    {
      "DeviceIndex": 0,
      "SubnetId": "subnet-1a2b3c4d",
      "Groups": [ "sg-1a2b3c4d" ],
      "AssociatePublicIpAddress": true
    }
  ],
  "IamInstanceProfile": {
      "Arn": "arn:aws:iam::123456789012:instance-profile/my-iam-role"
  }
}
```
### Terraform Version
- 0.7.7
### Affected Resource(s)
- aws_spot_instance_request
### Terraform Configuration Files

``` hcl
// OK (ondemand instance in vpc)
resource "aws_instance" "ondemand" {
  // public ip is associated.
  associate_public_ip_address = true

  user_data = "..."

  instance_type = "..."
  key_name = "..."
  subnet_id = "..."
  vpc_security_group_ids = [
    "..."
  ]
  tags {
    Name = "..."
  }
}

// NG (spot instance in vpc)
resource "aws_spot_instance_request" "spot" {
  // public ip is never associated
  associate_public_ip_address = true

  spot_price = "0.5"
  wait_for_fulfillment = true

  user_data = "..."

  ami = "..."
  instance_type = "..."
  key_name = "..."

  subnet_id = "..."
  vpc_security_group_ids = [
    "..."
  ]
  tags {
    Name = "..."
  }
}
```
### Expected Behavior

When "associate_public_ip_address=true", aws_spot_instance_request should associate public ip.
### Actual Behavior

aws_spot_instance_request never associates public ip.

**Comment by stack72** [comment]
Hi @maskinoshita 

Thanks for opening the issue here - sorry this is causing you some problems. Can you tell me if the subnet that you are launching the instance into actually allows public_ips to be mapped?

Paul

**Comment by davidski** [comment]
Forgive me for jumping in on this thread -- this has bit me as well. :smile:

Per the AWS docs (http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/vpc-ip-addressing.html), setting a subnet's public_ip auto-assign value to false has no effect if the EC2 reservation request has the public_ip value set to true. When using the spot_instance resource, I've been seeing reservations without the `AssociatePublicIpAddress`  set to true per http://docs.aws.amazon.com/cli/latest/reference/ec2/describe-spot-instance-requests.html.

Looking at the code, the resource_aws_spot_instance seems to have different logic (https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_spot_instance_request.go#L112-L114) than the on-demand version (https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_spot_instance_request.go#L112-L114), though I think the AWS-side structures are pretty much the same. Might this be the problem, @stack72 ?

**Comment by maskinoshita** [comment]
Hi, @stack72, @davidki

I wrote a reproduction code. Please see a following gist:
https://gist.github.com/maskinoshita/a7c439bfe9f2c519d45cc46946d9f19f

As @davidki said, ondemand and spot have different behaviors when subnet's map_public_ip_on_launch is set false.

I'm confused in this behavior.

**Comment by davidski** [comment]
Hey @stack72 - This issue still has the `waiting-response` label on it. Is there anything else you need to help get this into the queue? Happy to help out where I can, short of the Go coding. :smile:

**Comment by hrach** [comment]
Bug is still present in 0.9.2.

**Comment by hrach** [comment]
PR with fix is here if you would like to try: #13736
